### PR TITLE
Factor out a common parent interface for StatelessWorkflow and Workflow (now StatefulWorkflow).

### LINKED ISSUE #213: WorkflowContext.compose shouldn't have to know child workflow's state type.
This was only required back when we internally keyed off of the tuple of type parameters. Since we use the actual `KClass` of `Workflow` implementations now, we don't need to care anymore.

### PR DESCRIPTION
**Pull request by zach-klippenstein**
tl;dr:
 - `Workflow` -> `StatefulWorkflow`
 - `Workflow` is an interface, `StatefulWorkflow` and `StatelessWorkflow` are abstract classes
   that both implement `Workflow`.

This change makes it much easier for consumers of this library to define public APIs with private
implementations. The public API is an interface that extends `Workflow`. The private
implementation is a class that subclasses either `StatefulWorkflow` or `StatelessWorkflow`,
depending on whether it needs to keep track of any state. Either way, the workflow's state type
is kept as a completely private concern.

`interface Workflow<I, O, R>`
-----------------------------

This is the main currency of this library. `Workflow`s can compose
other `Workflow`s. A `Workflow` is anything that can be expressed as a `StatefulWorkflow`.
It has a single method, `asStatefulWorkflow()`, that is responsible for expressing the
workflow in terms of a `StatefulWorkflow` – a workflow that has an internal state. This
commit renames what used to be `Workflow` to `StatefulWorkflow`.

`abstract class StatefulWorkflow<I, S, O, R>`
---------------------------------------------

This is what you subclass to write a `Workflow` that has internal state that persists
as long as the workflow is active in its parent. It knows about things like initial state
and snapshotting/restoring.

`abstract class StatelessWorkflow<I, O, R>`
-------------------------------------------

This is what you subclass to get a `Workflow` that only has a single `compose` method.
It can compose children workflows, transform inputs/outputs/renderings up/down the tree,
and listen to external events and subscriptions, but it has no state.

Closes #213.

Based on #214 

[Internal discussion](https://square.slack.com/archives/CE470RL5T/p1553704080224400)

### CODE REVIEW

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt [hunk 1]
```diff
@@ -17,14 +17,43 @@ package com.squareup.workflow
 
 /**
  * A [Workflow] that has no state.
+ *
+ * @see StatefulWorkflow
  */
-typealias StatelessWorkflow<InputT, OutputT, RenderingT> =
-    Workflow<InputT, Unit, OutputT, RenderingT>
+interface StatelessWorkflow<InputT : Any, OutputT : Any, RenderingT : Any> :
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: I don't want to increase the scope of this PR, more a discussion of possible follow up work: I still find myself wondering if the code would be simpler (both for us to maintain and for others to consume) if `StatelessWorkflow` and `StatefulWorkflow` were abstract classes.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+ * to its parent.
+ * May be [Nothing] if the workflow doesn't need to emit anything.
+ *
+ * @param RenderingT The value returned to this workflow's parent during [composition][compose].
+ * Typically represents a "view" of this workflow's input, current state, and children's renderings.
+ * A workflow that represents a UI component may use a view model as its rendering type.
+ *
+ * @see StatelessWorkflow
+ */
+interface StatefulWorkflow<in InputT : Any, StateT : Any, out OutputT : Any, out RenderingT : Any> :
```
[the remark this answers is not in the document]
Comment (rjrjr) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: Pretty sure I'm kidding about that, but not completely.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt (hunk 1 above)
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]:
```kotlin
sealed class Workflow<I, O, R> {
  internal abstract fun asStatefulWorkflow(): StatefulWorkflow<I, *, O, R>
}

abstract class StatefulWorkflow<I, S, O, R> : Workflow<I, O, R>() {
  final override internal fun asStatefulWorkflow() = this

  abstract fun initialState(…
}

abstract class StatelessWorkflow<I, O, R> : Workflow<I, O, R>() {
  final override internal fun asStatefulWorkflow() = object : …
  
  abstract fun compose(I, …
}
```

Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: So using abstract classes (without a sealed parent) gives us the ability to make the `asStatefulWorkflow` override final in the 2 children, which would be nice.

Making them _sealed_ lets us get rid of that function entirely. We could have a separate `WorkflowNode` just for stateless workflows, which would be much simpler than the stateful one and could use less memory.

Those are the only two advantages of using abstract classes over interfaces that I can think of. Did you have anything else in mind @rjrjr?

Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: Trying to think if there are any disadvantages to using abstract classes…

The only thing is that it limits multiple inheritance, but I'm not sure that matters. In fact, it's probably even a nice benefit since other classes shouldn't be implementing workflow types anyway.

Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: Nvm about sealed classes - `Workflow` can't be a class because it would prevent consumers from using the public-interface-private-impl pattern (can't extend both `AuthWorkflow` and `StatefulWorkflow` if they're both classes).

Comment (rjrjr) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: So excited!

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: 🤔

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt (hunk 1 above)
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: This PR now also makes `StatelessWorkflow` and `StatefulWorkflow` abstract classes!

**Comment by zach-klippenstein** [comment]
Fleshed out the docs, added a commit message, this is now ready for real review. When it's ready, I'll squash the first commit out and decline #214.

cc @rjrjr

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow
+
+/**
+ * A composable, stateful object that can [handle events][WorkflowContext.onEvent],
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: The docs of the base classes and the interface seem unnecessarily redundant. I don't think a big deal for this PR, but something I'm planning to make a pass on once I'm done with the samples — heads up.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+ */
+abstract class StatefulWorkflow<
+    in InputT : Any,
+    StateT : Any,
+    out OutputT : Any,
+    out RenderingT : Any
+    > : Workflow<InputT, OutputT, RenderingT> {
+
+  /**
+   * Called when the state machine is first started to get the initial state.
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: Called from `WorkflowContext.compose` when a state machine is first started…

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+   */
+  abstract fun initialState(input: InputT): StateT
+
+  /**
+   * Called whenever [WorkflowContext.compose] is about to get a new [input][InputT] value, to allow
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: Called from `WorkflowContext.compose` instead of `initialState` when a workflow is already running. This 
allows the workflow to detect changes in input, and possibly change its state in response.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    old: InputT,
+    new: InputT,
+    state: StateT
+  ): StateT = state
+
+  /**
+   * Called at least once† any time one of the following things happens:
+   *  - This workflow's [input] changes (via the parent passing a different one in).
+   *  - This workflow's [state] changes.
+   *  - A child workflow's state changes.
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: Instead of _A child workflow's state changes_, how about _A child workflow emits output, or updates its rendering._

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+   *
+   * **Never call this method directly.** To get the rendering from a child workflow, pass the child
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: To nest the rendering of a child workflow in your own,

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+  /**
+   * Deserialize a state value from a [Snapshot] previously created with [snapshotState].
+   *
+   * If the workflow should always be started from scratch, this method can just ignore the snapshot
+   * and return the initial state.
+   *
+   * @see snapshotState
+   */
+  abstract fun restoreState(snapshot: Snapshot): StateT
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: See #220

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt [hunk 9]
```diff
@@ -16,15 +16,78 @@
...
  */
-typealias StatelessWorkflow<InputT, OutputT, RenderingT> =
-    Workflow<InputT, Unit, OutputT, RenderingT>
+abstract class StatelessWorkflow<InputT : Any, OutputT : Any, RenderingT : Any> :
+    Workflow<InputT, OutputT, RenderingT> {
+
+  /**
+   * Called at least once any time one of the following things happens:
+   *  - This workflow's [input] changes (via the parent passing a different one in).
+   *  - A child workflow updates in some way (e.g. a stateful child's state changes).
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: s/_(e.g.…)/(i.e., it emits output or updates its rendering)/

And add a bullet about workers?

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/Workflow.kt [hunk 10]
```diff
@@ -18,107 +18,55 @@ package com.squareup.workflow
...
-  /**
-   * Deserialize a state value from a [Snapshot] previously created with [snapshotState].
-   *
-   * If the workflow should always be started from scratch, this method can just ignore the snapshot
-   * and return the initial state.
-   *
-   * @see snapshotState
-   */
-  fun restoreState(snapshot: Snapshot): StateT
+  fun asStatefulWorkflow(): StatefulWorkflow<InputT, *, OutputT, RenderingT>
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/Workflow.kt]: Provides a `StatefulWorkflow` view of the receiver. Necessary because `StatefulWorkflow` is the common API required for `WorkflowContext.compose()` to do its work.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/Workflows.kt [hunk 11]
```diff
@@ -24,26 +24,10 @@ import com.squareup.workflow.WorkflowAction.Companion.emitOutput
...
       transform: (FromRenderingT) -> ToRenderingT
-    ): StatelessWorkflow<InputT, OutputT, ToRenderingT> = StatelessWorkflow { input, context ->
+    ): Workflow<InputT, OutputT, ToRenderingT> = StatelessWorkflow { input, context ->
 // @formatter:on
   context.compose(this@mapRendering, input) { emitOutput(it) }
       .let(transform)
 }
-
-/**
- * Returns a rendering that directly delegates to this [Workflow] but masks its state type.
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/Workflows.kt]: 🔥

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt (hunk 3 above)
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: Yea, I'm not a fan either. Probably fine to put most of the docs on `Workflow`, point both children to it, and then only have state details on `StatefulWorkflow`.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt (hunk 6 above)
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: But state changes _do_ trigger composition, and rendering changes don't (because "render" is just another word for this method). Updating to include both state and output at any rate.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt (hunk 9 above)
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: What would you like to see about workers? That term doesn't exist in any of the v2 code yet. I've updated this comment to match the updated wording for `StatefulWorkflow`.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/Workflow.kt (hunk 10 above)
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/Workflow.kt]: And this is why i always leave `TODO kdoc` comments everywhere…

**Comment by zach-klippenstein** [comment]
Thanks for the merge conflict fixes @rjrjr 😁 Gonna merge when CI goes green.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt (hunk 6 above)
Comment (rjrjr) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt]: Ah, right right.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt (hunk 9 above)
Comment (rjrjr) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/StatelessWorkflow.kt]: _ That term doesn't exist in any of the v2 code yet_ Excellent.

**Comment by zach-klippenstein** [comment]
Travis is having issues:
image

I ran `./gradlew clean connectedCheck` locally, everything passed, so I'm just gonna merge.
